File: examples/cpp/tutorials/src/example6_execute_vla.cpp
#include<iostream>
#include<fstream>
#include<thread>
#include<chrono>
#include "opencv2/opencv.hpp"

#include "galbot_robot.hpp"
#include "galbot_motion.hpp"
#include "galbot_navigation.hpp"

using namespace galbot::sdk::g1;

/**
 * Generate target point for trajectory
 */
TrajectoryPoint generateTargetPoint(const std::vector<double>& q, double targetTime = 10.0) {
    TrajectoryPoint jointPosition;
    jointPosition.time_from_start_second = targetTime;
    
    for (double joint : q) {
        JointCommand jointCmd;
        jointCmd.position = joint;
        jointPosition.joint_command_vec.push_back(jointCmd);
    }
    
    return jointPosition;
}

/**
 * Generate trajectory for joints
 */
std::shared_ptr<Trajectory> generateTargetTrajectory(
    const std::vector<std::vector<double>>& trajectory,
    const std::vector<std::string>& jointGroups,
    const std::vector<std::string>& jointNames,
    double dt = 0.008
) {
    if (trajectory.empty()) {
        return nullptr;
    }
    
    auto traj = std::make_shared<Trajectory>();
    traj->joint_groups = jointGroups;
    traj->joint_names = jointNames;
    
    double currentTime = 0.0;
    for (const auto& state : trajectory) {
        currentTime += dt;
        TrajectoryPoint trajPoint = generateTargetPoint(state, currentTime);
        traj->points.push_back(trajPoint);
    }
    
    return traj;
}

/**
 * Print joint states
 */
void printJointStates(const std::vector<JointState>& jointStates) {
    for (const auto& js : jointStates) {
        std::cout << " : position = " << js.position 
             << " , velocity = " << js.velocity
             << " , acceleration = " << js.acceleration
             << " , effort = " << js.effort
             << " , current = " << js.current << std::endl;
    }
}

/**
 * Fake VLA (Vision-Language-Action) model implementation
 */
std::map<std::string, std::vector<std::vector<double>>> fakeVla(const std::unordered_map<SensorType, std::shared_ptr<cv::Mat>>& rgbDataDict) {
    std::cout << "Fake VLA executing..." << std::endl;
    
    const int numPoints = 200;
    std::map<std::string, std::vector<std::vector<double>>> result;
    
    // Right arm trajectory
    std::vector<double> rightArmStart = {-2.0, 1.59, 0.6, 1.7, 0.0, 0.8, 0.0};
    std::vector<double> rightArmEnd = {-1.5, 1.59, 0.6, 1.5, 0.0, 0.6, 0.0};
    std::vector<std::vector<double>> rightArmTraj(numPoints);
    for (int i = 0; i < numPoints; ++i) {
        std::vector<double> point(7);
        for (int j = 0; j < 7; ++j) {
            point[j] = rightArmStart[j] + (rightArmEnd[j] - rightArmStart[j]) * i / (numPoints - 1);
        }
        rightArmTraj[i] = point;
    }
    
    // Left arm trajectory
    std::vector<double> leftArmPos = {1.9999, -1.6000, -0.5999, -1.6999, 0.0000, -0.7999, 0.0000};
    std::vector<std::vector<double>> leftArmTraj(numPoints, leftArmPos);
    
    // Head trajectory
    std::vector<double> headPos = {0.0, 0.0};
    std::vector<std::vector<double>> headTraj(numPoints, headPos);
    
    // Leg trajectory
    std::vector<double> legPos = {0.299, 1.199, 0.849, 0.0000, 0.0};
    std::vector<std::vector<double>> legTraj(numPoints, legPos);
    
    result["leg"] = legTraj;
    result["head"] = headTraj;
    result["left_arm"] = leftArmTraj;
    result["right_arm"] = rightArmTraj;
    
    return result;
}

std::map<std::string, std::vector<std::vector<double>>> estimate_vla(GalbotRobot& robot, const std::unordered_set<SensorType>& enable_sensor_set){
    /** Get RGB images from enabled cameras */
    std::unordered_map<SensorType, std::shared_ptr<cv::Mat>> rgb_images;
    for(const auto& sensor_type : enable_sensor_set){
        if(sensor_type == SensorType::LEFT_ARM_CAMERA || sensor_type == SensorType::RIGHT_ARM_CAMERA ||
            sensor_type == SensorType::HEAD_LEFT_CAMERA || sensor_type == SensorType::HEAD_RIGHT_CAMERA){
            std::shared_ptr<RgbData> rgb_data = robot.get_rgb_data(sensor_type);
            if(rgb_data){
                std::shared_ptr<cv::Mat> rgb_image = rgb_data->convert_to_cv2_mat();
                rgb_images[sensor_type] = rgb_image;
            }else{
                std::cerr << "Failed to get RGB data. " << std::endl;
            }
        }
    }
    std::cout << "RGB images size: " << rgb_images.size() << std::endl;

     /** Estimate VLA */
    std::map<std::string, std::vector<std::vector<double>>> vla = fakeVla(rgb_images);
    return vla;
}

/* @brief Check if the robot is safe
*/
void check_robot_safety(){
    std::cout << "⚠️  Note: 1. Please ensure the robot's emergency stop button is released; 2. Please ensure there are no obstacles in front, back, left, and right of the robot to avoid unexpected situations. \n" << std::endl;

    char key;
    for(;;){
        std::cout << "Please confirm that the robot's emergency stop button is released and there are no obstacles. Continue? (y/n)...";
        std::cin >> key;

        if(std::tolower(key) == 'y'){
            std::cout << "User confirmed, continuing execution...\n" << std::endl;
            break;
        }else if(std::tolower(key) == 'n'){
            std::cout << "User not confirmed, program exiting...\n" << std::endl;
            exit(0);
        }else{
            std::cout << "Input error, please enter 'y' or 'n'\n" << std::endl;
        }
    }
}

int main(){
    check_robot_safety();
    try{
        /* Get robot instance  */
        GalbotRobot& robot = GalbotRobot::get_instance();
        GalbotMotion& motion = GalbotMotion::get_instance();
        GalbotNavigation& navigation = GalbotNavigation::get_instance();

        /** Enable sensor type */
        std::unordered_set<SensorType> enable_sensor_set = {
            SensorType::LEFT_ARM_CAMERA,
            SensorType::RIGHT_ARM_CAMERA,
            SensorType::HEAD_LEFT_CAMERA,
            SensorType::HEAD_RIGHT_CAMERA
        };

        /* Initialize robot */
        if (robot.init(enable_sensor_set)) {
            std::cout << "Initialization successful" << std::endl;
            std::cout << "Is robot running: " << robot.is_running() << std::endl;
        }else{
            std::cerr << "Initialization failed" << std::endl;
        }
        if (!motion.init()) {
            std::cerr << "Motion initialization failed" << std::endl;
        }else{
            std::cout << "Motion initialization successful" << std::endl;
        }
        if (!navigation.init()) {
            std::cerr << "Navigation initialization failed" << std::endl;
        }else{
            std::cout << "Navigation initialization successful" << std::endl;
        }

        /* Wait for data preparation */
        std::this_thread::sleep_for(std::chrono::milliseconds(3000));
        
        /** Estimate VLA actions */
        std::map<std::string, std::vector<std::vector<double>>> joint_positions = estimate_vla(robot, enable_sensor_set);

        // Generate target trajectory
        std::vector<std::string> jointGroups;
        std::vector<std::vector<double>> jointTraj;
        
        for (const auto& [group, traj] : joint_positions) {
            jointGroups.push_back(group);
            if (jointTraj.empty()) {
                jointTraj = traj;
            } else {
                // Concatenate trajectories
                for (size_t i = 0; i < traj.size(); ++i) {
                    jointTraj[i].insert(jointTraj[i].end(), traj[i].begin(), traj[i].end());
                }
            }
        }
        
        // Final joint position check
        std::vector<double> wholeBodyJoint = jointTraj.back();
        std::vector<double> baseState = {0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0};
        
        RobotStates checkState;
        checkState.whole_body_joint = wholeBodyJoint;
        checkState.base_state = baseState;
        
        std::cout << "✅ Final joint position check state: [";
        for (double val : wholeBodyJoint) std::cout << val << " ";
        std::cout << "]" << std::endl;
        
        // Check collision
        // std::tuple<MotionStatus, std::vector<bool>> collisionResult = motion.check_collision(std::vector<std::shared_ptr<RobotStates>>{std::make_shared<RobotStates>(checkState)}, true);
        // MotionStatus status = std::get<0>(collisionResult);
        // std::vector<bool> collisionRes = std::get<1>(collisionResult);
        auto [status, collisionRes] = motion.check_collision(std::vector<std::shared_ptr<RobotStates>>{std::make_shared<RobotStates>(checkState)}, true);
        std::this_thread::sleep_for(std::chrono::seconds(1));
        
        if (status == MotionStatus::SUCCESS) {
            std::cout << "✅ OK: collision check finished: " << collisionRes[0] 
                 << " (False=no collision)" << std::endl;
            
            // Execute trajectory
            auto trajectory = generateTargetTrajectory(jointTraj, jointGroups, {});
            if (trajectory != nullptr) {
                ControlStatus execStatus = robot.execute_joint_trajectory(*trajectory, true);
                std::this_thread::sleep_for(std::chrono::seconds(1));
                
                if (execStatus == ControlStatus::SUCCESS) {
                    std::cout << "✅ Joint trajectory execution successful." << std::endl;
                } else {
                    std::cout << "❌ Joint trajectory execution failed." << std::endl;
                }
                
                // Check joint state
                auto jointStates = robot.get_joint_states(jointGroups, {});
                printJointStates(jointStates);
                std::cout << "✅ Final joint position check state after execution" << std::endl;
            } else {
                std::cout << "❌ Generated trajectory is invalid, cannot execute." << std::endl;
            }
        } else {
            std::cout << "❌ Collision check failed, will not execute the joint trajectory." << std::endl;
        }

        /** Actively send SIGINT exit signal to the robot */
        robot.request_shutdown();
        /** Wait to enter shutdown state */
        robot.wait_for_shutdown();
        /** Release SDK resources */
        robot.destroy();
        std::cout << "Resource release successful" << std::endl;    
    }catch(const std::exception& e){
        std::cout << "Error: " << e.what() << std::endl;
    }

    return 0;
}

File: docs/g1/zh/python/tutorials/example6_execute_vla.py
"""
Note: When running this example, please ensure the robot's `emergency stop button` is released;
"""
import time
import os
from typing import List, Dict, Any

try:
    import galbot_sdk.g1 as gm
    from galbot_sdk.g1 import GalbotRobot
    from galbot_sdk.g1 import GalbotMotion
    from galbot_sdk.g1 import GalbotNavigation
    from galbot_sdk.g1 import (
        SensorType, G1JointGroup, ControlStatus,
        Trajectory, TrajectoryPoint, JointCommand
    )
except ImportError:
    print("Failed to import galbot_sdk, please install it first or check if it is in PYTHONPATH")
    exit(1)

try:
    import cv2
except ImportError:
    os.system("pip install opencv-python")
    import cv2

try:
    import numpy as np
except ImportError:
    os.system("pip install numpy")
    import numpy as np


def decode_compressed_image(compressed_image: Dict[str, Any]) -> np.ndarray:
    """
    Decode CompressedImage image.

    Parameters:
        compressed_image (dict): image dict, keys: [header, format, data, "depth_scale"]

    Returns:
        numpy.ndarray: decoded image
    """
    image_data = compressed_image["data"]
    if compressed_image["format"] == "rgb8":
        return decode_rgb_image(image_data)
    elif compressed_image["format"] == "16UC1":
        return decode_depth_image(image_data)
    else:
        raise ValueError(f"Unsupported data format: {compressed_image['format']}")


def decode_rgb_image(image_data: bytes) -> np.ndarray:
    """
    Decode RGB image.

    Parameters:
        image_data (bytes): Raw image data

    Returns:
        numpy.ndarray: Decoded RGB image
    """
    nparr = np.frombuffer(image_data, np.uint8)
    img = cv2.imdecode(nparr, cv2.IMREAD_COLOR)
    if img is None:
        raise ValueError("Failed to decode RGB Image")
    return img


def decode_depth_image(image_data: Dict[str, Any]) -> np.ndarray:
    """
    Decode depth image.

    Parameters:
        image_data (dict): Depth image data with metadata

    Returns:
        numpy.ndarray: Decoded depth image
    """
    depth_img = np.frombuffer(image_data["data"], dtype=np.uint16).copy()

    # Check if height and width exist
    if "height" not in image_data or "width" not in image_data:
        raise ValueError("Missing 'height' or 'width' in depth image metadata.")
    if image_data["height"] == 0 or image_data["width"] == 0:
        raise ValueError(
            f"Invalid 'height' ({image_data['height']}) or 'width' ({image_data['width']}) in depth image metadata."
        )

    # Parse depth image
    depth_img = depth_img.reshape((image_data["height"], image_data["width"]))
    depth_img = depth_img.astype(np.float32) / image_data["depth_scale"]

    return depth_img


def fake_vla(rgb_data_dict: dict) -> Dict[str, np.ndarray]:
    """
    Fake VLA (Vision-Language-Action) model implementation.
    
    This function is a placeholder. In a real-world scenario, you would implement
    target detection using computer vision techniques. For this example, we assume
    a default pose.

    Parameters:
        rgb_data_dict (dict): Dictionary containing RGB image data from various sensors

    Returns:
        dict: Trajectories for different robot components (legs, head, arms)
    """
    print("Fake VLA executing...")
    
    # Define trajectories for different robot parts
    right_arm_traj = np.linspace(
        [-2.0, 1.59, 0.6, 1.7, 0.0, 0.8, 0.0],
        [-1.5, 1.59, 0.6, 1.5, 0.0, 0.6, 0.0],
        num=200,
    )

    left_arm_traj = np.linspace(
        [1.9999, -1.6000, -0.5999, -1.6999, 0.0000, -0.7999, 0.0000],
        [1.9999, -1.6000, -0.5999, -1.6999, 0.0000, -0.7999, 0.0000],
        num=200,
    )
    
    head_traj = np.linspace(
        [0.0, 0.0],
        [0.0, 0.0],
        num=200,
    )

    leg_traj = np.linspace(
        [0.299, 1.199, 0.849, 0.0000, 0.0],
        [0.299, 1.199, 0.849, 0.0000, 0.0],
        num=200,
    )

    return {
        "leg": leg_traj,
        "head": head_traj,
        "left_arm": left_arm_traj,
        "right_arm": right_arm_traj,
    }


def estimate_vla(robot: 'GalbotRobot', enable_sensor_set: set) -> Dict[str, np.ndarray]:
    """
    Estimate VLA (Vision-Language-Action) model outputs based on sensor data.

    Parameters:
        robot (GalbotRobot): GalbotRobot instance
        enable_sensor_set (set): Set of enabled sensor types

    Returns:
        dict: Joint positions for each joint group
    """
    # Get RGB images from enabled cameras
    rgb_data_dict = {}
    for sensor_type in enable_sensor_set:
        if sensor_type in [
            SensorType.LEFT_ARM_CAMERA, 
            SensorType.RIGHT_ARM_CAMERA, 
            SensorType.HEAD_LEFT_CAMERA, 
            SensorType.HEAD_RIGHT_CAMERA
        ]:
            # Get RGB data of sensor type
            rgb_data = robot.get_rgb_data(sensor_type)
            time.sleep(1)
            
            # Decode RGB image
            if rgb_data:
                rgb_image = decode_compressed_image(rgb_data)
                rgb_data_dict[sensor_type] = rgb_image
            else:
                print(f"Failed to get RGB data from {sensor_type}")
        else:
            print(f"Unsupported sensor type: {sensor_type}")
    
    print("RGB data dictionary keys:", rgb_data_dict.keys())

    # Get joint position list
    joint_positions = fake_vla(rgb_data_dict)
    return joint_positions


def generate_target_point(q: List[float], target_time: float = 10.0) -> 'TrajectoryPoint':
    """
    Generate target points for trajectory execution.

    Parameters:
        q (List[float]): Joint positions
        target_time (float): Time from start in seconds (default: 10.0)

    Returns:
        TrajectoryPoint: Target trajectory point with specified joint commands
    """
    joint_position = TrajectoryPoint()
    joint_position.time_from_start_second = target_time
    joint_command_vec = []
    for joint in q:
        joint_cmd = JointCommand()
        joint_cmd.position = joint
        joint_command_vec.append(joint_cmd)
    joint_position.joint_command_vec = joint_command_vec
    return joint_position


def generate_target_trajectory(
    trajectory: np.ndarray, 
    joint_groups: List[str] = None, 
    joint_names: List[str] = None, 
    dt: float = 0.008
) -> 'Trajectory':
    """
    Generate trajectory for joints.

    Parameters:
        trajectory (np.ndarray): 2D array of joint positions over time
        joint_groups (List[str]): List of joint groups
        joint_names (List[str]): List of joint names
        dt (float): Time step between trajectory points (default: 0.008)

    Returns:
        Trajectory: Generated trajectory for execution
    """
    if joint_groups is None:
        joint_groups = []
    if joint_names is None:
        joint_names = []

    if trajectory is None or np.ndim(trajectory) != 2 or len(trajectory) == 0:
        return None

    # Create Trajectory
    traj = Trajectory()
    traj.joint_groups = joint_groups
    traj.joint_names = joint_names

    current_time = 0.0
    points = []
    for state in trajectory:
        current_time += dt
        # Generate target point for each joint
        traj_point = generate_target_point(state, current_time)
        points.append(traj_point)

    traj.points = points
    return traj


def print_joint_states(joint_states: List['JointState']) -> None:
    """
    Print joint states in a readable format.

    Parameters:
        joint_states (List[JointState]): List of joint states
    """
    for js in joint_states:
        print(
            f" : position = {js.position} , velocity = {js.velocity} "
            f", acceleration = {js.acceleration} , effort = {js.effort} , current = {js.current}"
        )


def check_robot_safety() -> None:
    """
    Check if the robot is safe to operate.
    Prompts the user to confirm safety conditions before proceeding.
    """
    # Prompt for precautions
    print(
        "⚠️  Note: 1. Please ensure the robot's emergency stop button is released; "
        "2. Please ensure there are no obstacles in front, back, left, and right "
        "of the robot to avoid unexpected situations."
    )
    
    while True:
        key = input(
            "Please confirm that the robot's emergency stop button is released "
            "and there are no obstacles. Continue? (y/n)..."
        ).lower()
        
        if key == 'y':
            print("User confirmed, continuing execution...")
            break
        elif key == 'n':
            print("User not confirmed, program exiting...")
            exit(1)
        else:
            print("Input error, please enter 'y' or 'n'")


def main() -> None:
    """
    Main function to execute the VLA (Vision-Language-Action) example.
    Initializes the robot, estimates actions based on sensor data, and executes trajectories.
    """
    check_robot_safety()

    robot = None  # Initialize robot variable for cleanup in finally block

    try:
        # Get robot instances
        robot = GalbotRobot()
        motion = GalbotMotion()
        navi = GalbotNavigation()

        # Enable required sensors
        enable_sensor_set = {
            SensorType.LEFT_ARM_CAMERA,
            SensorType.RIGHT_ARM_CAMERA,
            SensorType.HEAD_LEFT_CAMERA,
            SensorType.HEAD_RIGHT_CAMERA
        }
                             
        # Initialize robot components
        if not robot.init(enable_sensor_set):
            print("Robot initialization failed")
            exit(1)
        else:
            print("Robot initialization successful")
            
        if not motion.init():
            print("Motion initialization failed")
            exit(1)
        else:
            print("Motion initialization successful")
            
        if not navi.init():
            print("Navigation initialization failed")
            exit(1)
        else:
            print("Navigation initialization successful")
        
        # Wait for data preparation
        time.sleep(3)

        # Estimate VLA actions
        joint_positions = estimate_vla(robot, enable_sensor_set)
        
        # Generate target trajectory
        joint_groups = list(joint_positions.keys())
        joint_traj = np.concatenate(list(joint_positions.values()), axis=-1)
        
        # Final joint position check state
        whole_body_joint = joint_traj[-1]
        base_state = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 1.0]
        check_state = gm.RobotStates()
        check_state.whole_body_joint = whole_body_joint
        check_state.base_state = base_state
        print(f"✅ Final joint position check state: {whole_body_joint}")
        
        # Check collision
        status, collision_res = motion.check_collision(
            start=[check_state],
            enable_collision_check=True
        )
        time.sleep(1)

        if status == gm.MotionStatus.SUCCESS:
            print(f"✅ OK: collision check finished: {collision_res} (False=no collision)")

            # Execute trajectory
            trajectory = generate_target_trajectory(joint_traj, joint_groups, [])
            if trajectory is not None:
                status = robot.execute_joint_trajectory(trajectory, True)
                time.sleep(1)
                
                if status == ControlStatus.SUCCESS:
                    print("✅ Joint trajectory execution successful.")
                else:
                    print("❌ Joint trajectory execution failed.")
                
                # Check joint state
                joint_states = robot.get_joint_states(joint_groups, [])
                print_joint_states(joint_states)
                final_positions = [getattr(js, "position", js) for js in joint_states]
                print(f"✅ Final joint position check state after execution (positions): {final_positions}")
            else:
                print("❌ Generated trajectory is invalid, cannot execute.")
        else:
            print("❌ Collision check failed, will not execute the joint trajectory.")

    except Exception as e:
        print(f"An exception occurred: {e}")
        import traceback
        traceback.print_exc()
    finally:
        # Ensure proper cleanup
        if robot is not None:
            # Actively send SIGINT shutdown signal
            robot.request_shutdown()
            # Wait to enter shutdown state
            robot.wait_for_shutdown()
            # Release SDK resources
            robot.destroy()
        print('Resource release successful')


if __name__ == "__main__":
    main()